New How-To chapter (#100)
pyansys/dev-guide: doc/source/how-to/setup.rst

Development Environment

> =======================

Author: Thanks for pointing this out, @PipKat. I was trying to play with the headings to get a useful slug:

```
https://.../how-to/setup/#development-environment
```

In fact, should we rename this file like `setup.rst -> setting-up.rst` so we can get:

```
https://.../how-to/setting-up/#development-environment
```

?

> Before starting contributing to any ``PyAnsys`` projects, you will need to setup
> your developers environment.

Reviewer: ```suggestion
Before you can contribute to any ``PyAnsys`` project, you must set up
your developer environment.
```

Python
------
.. raw:: html
    
    <div align="center">
      <img src="https://github.com/python/pythondotorg/raw/main/static/community_logos/python-logo-inkscape.svg">
    </div>

> All ``PyAnsys`` projects require from a Python interpreter in order to interact
> with the software. Therefore, you want to make sure that at least one Python
> interpreter is installed in your local machine.

Reviewer: ```suggestion
All ``PyAnsys`` projects require a Python interpreter for interacting
with the software. Therefore, you must ensure that at least one Python
interpreter is installed on your local machine.
```

> Installation
> ~~~~~~~~~~~~

Reviewer: ```suggestion
Installing Python
~~~~~~~~~~~~~~
```

> There are multiple ways to install Python in your local machine:

Reviewer: ```suggestion
There are multiple ways to install a Python package on your local machine:
```

> - Using an official installer from the `official Python download section <https://www.python.org/downloads/>`_.
> - Installing it via a package manager or "store" in your machine.

Reviewer: ```suggestion
- Use an official installer from the `official Python download section <https://www.python.org/downloads/>`_.
- Install via a package manager or "store" on your machine.
```

.. warning:: 

>    Please, make sure you install Python from an official channel. Do not trust
>    third party websites nor download executable content from those.

Reviewer: ```suggestion
  Ensure that you install Python from an official channel. Do not trust
   third-party websites or download executable content from them.
```

.. tabs::

    .. group-tab:: Windows

>         To install Python in a machine using Windows, follow these steps:

Reviewer: ```suggestion
        To install Python on a machine running Windows:
```

        
        1. Download the `latest stable Python version for Windows <https://www.python.org/downloads/windows/>`_.
        2. Execute the installer and follow the different steps.

>         For a more detailed installation process, please check `Using Python on
>         Windows <https://docs.python.org/3/using/windows.html>`_.

Reviewer: ```suggestion
        For detailed installation instructions, see `Using Python on
        Windows <https://docs.python.org/3/using/windows.html>`_.
```

>         2. Execute the installer and follow the different steps.
> 
>         For a more detailed installation process, please check `Using Python on
>         Windows <https://docs.python.org/3/using/windows.html>`_.

Reviewer: ```suggestion
        2. Execute the installer, referring to  `Using Python on
        Windows <https://docs.python.org/3/using/windows.html>`_ for
        detailed installation information.
```

    .. group-tab:: macOS

>         To install Python in a machine using macOS, follow these steps:

Reviewer: ```suggestion
        To install Python on a machine running the macOS:
```

        
        1. Download the `latest stable Python version for macOS <https://www.python.org/downloads/macos/>`_.

>         2. Execute the installer and follow the different steps.
> 
>         For a more detailed installation process, please check `Using Python on
>         a Mac <https://docs.python.org/3/using/mac.html>`_.

Reviewer: ```suggestion
        2. Execute the installer, referring to `Using Python on
        a Mac <https://docs.python.org/3/using/mac.html>`_ for
        detailed installation information.
```

        .. note::

>             It is likely that your macOS distribution already comes with some
>             Python version installed. Please, refer to :ref:`Verify Installation`.

Reviewer: ```suggestion
            It is likely that your macOS distribution already comes with some
            version of Python installed. For more information, see :ref:`Verify Installation`.
```

    .. group-tab:: Linux/UNIX

>         To install Python in a machine using Linux/UNIX, follow these steps:

Reviewer: ```suggestion
        To install Python on a machine running Linux/UNIX:
```

        
        1. Download the `latest stable Python version for Linux/UNIX <https://www.python.org/downloads/source/>`_.

>         2. Decompress the source code and follow the installation instructions in README.rst.

Reviewer: ```suggestion
        2. Decompress the source code and follow the installation instructions in the ``README.rst`` file.
```

>         2. Decompress the source code and follow the installation instructions in README.rst.
> 
>         For a more detailed installation process, please check `Using Python on
>         Unix Platforms <https://docs.python.org/3/using/unix.html>`_.

Reviewer: ```suggestion
        2. Decompress the source code and follow the installation instructions in README.rst.
        For more detailed installation information, see `Using Python on
        Unix Platforms <https://docs.python.org/3/using/unix.html>`_.
```

        .. note::

>             It is likely that your Linux/UNIX distribution already comes with some
>             Python version installed. Please, refer to :ref:`Verify Installation`.

Reviewer: ```suggestion
            It is likely that your Linux/UNIX distribution already comes with some
            version of Python installed. For more information, see :ref:`Verify Installation`.
```

> Verify Installation
> ~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Verification
~~~~~~~~~~~
```

> Once your installation process is complete, verify your Python installation by
> running:

Reviewer: ```suggestion
Once your installation process is complete, verify your Python installation with:
```

.. tabs::

    .. group-tab:: Windows CMD

        .. code-block:: text

            py --version

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            py --version

    .. group-tab:: macOS

        .. code-block:: text

            python --version

    .. group-tab:: Linux/UNIX

        .. code-block:: text

            python --version

Virtual Environments
--------------------

> When working in multiple Python projects, it is likely each one of those has its
> own requirements. Sometimes, it may be possible that these requirements are
> incompatible across projects. To isolate Python environments and guarantee that
> no dependency problems are faced when workin with multiple projects, virtual
> environments were devised.
> 
> This subsection collects the most fundamental command for manipulating and
> interacting with Python virtual environments. For more information about these,
> please refer to `official Python documentation on venv module
> <https://docs.python.org/3/library/venv.html>`_.

Reviewer: ```suggestion
When working in multiple Python projects, it is likely each of these projects has its
own requirements. Sometimes, requirements across projects can be incompatible.
Virtual environments were devised to isolate Python environments, which guarantees
that you do not face dependency problems when working in multiple projects.

This subsection collects the most fundamental command for manipulating and
interacting with Python virtual environments. For more information about these,
please refer to `official Python documentation on venv module
<https://docs.python.org/3/library/venv.html>`_.
```

Check
~~~~~

> Before creating a new virtual environment, you need to check if you are already
> working with one. To do so, run the following command:

Reviewer: ```suggestion
Before creating a new virtual environment, you must check to see if you are already
working with one with this command:
```

.. tabs::

    .. group-tab:: Windows CMD

        .. code-block:: text

            where.exe python

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            where.exe python

    .. group-tab:: macOS

        .. code-block:: text

            which python

    .. group-tab:: Linux/UNIX

        .. code-block:: text

            which python

> Previous command will return the path to the currently used Python environment
> in your system. 
> 
> Make sure it points towards your default installation and not towards a virtual
> environment. If this is the case, then refer to :ref:`Deactivate` section to
> deactivate your virtual environment.

Reviewer: ```suggestion
This command will return the path to your system's currently used Python environment. 

Ensure that it points to your default installation and not to a virtual
environment. If it points to a virtual environment, see :ref:`Deactivate` for
information on deactivating your virtual environment.
```

Create
~~~~~~

> To create a virtual environment named `<venv>`, execute the following command:

Reviewer: ```suggestion
To create a virtual environment named `<venv>`, run this command:
```

.. tabs::

    .. group-tab:: Windows CMD

        .. code-block:: text

            py -m venv <venv>

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            py -m venv <venv>

    .. group-tab:: macOS

        .. code-block:: text

            python -m venv <venv>

    .. group-tab:: Linux/UNIX

        .. code-block:: text
            
            python -m venv <venv>

> Usually, virtual environments are named as ``venv`` or ``.venv``.

Reviewer: ```suggestion
Usually, virtual environments are named ``venv`` or ``.venv``.
```

Activate
~~~~~~~~

> To activate a virtual environment, execute the following command:

Reviewer: ```suggestion
To activate a virtual environment, run this command:
```

.. tabs::

    .. group-tab:: Windows CMD

        .. code-block:: text

            <venv>\Scripts\activate.bat

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            <venv>\Scripts\Activate.ps1

    .. group-tab:: macOS

        .. code-block:: text

            source <venv>/bin/activate

    .. group-tab:: Linux/UNIX

        .. code-block:: text

            source <venv>/bin/activate

Deactivate
~~~~~~~~~~

> 

Reviewer: ```suggestion
To deactivate a virtual environment, run this command:
```

.. tabs:: 

    .. group-tab:: Windows CMD

        .. code-block:: text

            deactivate

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            deactivate

    .. group-tab:: macOS

        .. code-block:: text

            deactivate

    .. group-tab:: Linux/UNIX

        .. code-block:: text

            deactivate

Git
---
.. raw:: html
    
    <div align="center">
      <img src="https://github.com/git/git-scm.com/raw/main/public/images/logo%402x.png">
      <br><br>
    </div>

> `Git <https://git-scm.com/>`_ is an open-source version control system (VCS). It
> is used to track changes and register new ones in software related projects. Git
> will register author and dates too, so accurate tracking of the evolution of the
> software can be made.

Reviewer: ```suggestion
`Git <https://git-scm.com/>`_ is an open-source version control system (VCS). It
is used to track changes and register new content in software-related projects. Git
will register the author and date of the changes so that accurate tracking of the
software's evolution is available.
```

> Installation

Reviewer: Is this installation information for installing the software needed in the virtual environment? I couldn't tell if that was the case or if this is redundant from the information given earlier.

~~~~~~~~~~~~

.. tabs::

    .. group-tab:: Windows

>         To install Python in a machine using Windows, follow these steps:
>         
>         1. Download the `latest stable standalone Git version for Windows <https://www.python.org/downloads/win/>`_.
>         2. Execute the installer and follow the different steps.

Reviewer: ```suggestion
        To install Python on a machine running Windows:
        
        1. Download the `latest stable standalone Git version for Windows <https://www.python.org/downloads/win/>`_.
        2. Execute the installer and follow the installation steps.
```

    .. group-tab:: macOS

>         To install Python in a machine using macOS, follow these steps:
>         
>         1. Check the `latest stable Git version for macOS <https://git-scm.com/download/mac>`_.
>         2. Run the installation command for your package manager.

Reviewer: ```suggestion
        To install Python on a machine running the macOS:
        
        1. Check the `latest stable Git version for macOS <https://git-scm.com/download/mac>`_.
        2. Run the installation command for your package manager.
```

    .. group-tab:: Linux/UNIX

>         To install Git in a machine using Linux/UNIX, follow these steps:

Reviewer: ```suggestion
        To install Git on a machine running Linux/UNIX:
```

        
        1. Check the `latest stable Git version for Linux/UNIX <https://git-scm.com/download/linux>`_.
        2. Run the installation command for your package manager.

> Verify Installation
> ~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Verification
~~~~~~~~~
```

> Check if the installation process succeeded by running the following command:

Reviewer: ```suggestion
Once your installation process is complete, verify your Python installation by
running:
```

.. tabs:: 

    .. group-tab:: Windows CMD

        .. code-block:: text

            git --version

    .. group-tab:: Windows PowerShell

        .. code-block:: text

            git --version

    .. group-tab:: macOS

        .. code-block:: text

            git --version

    .. group-tab:: Linux/UNIX

        .. code-block:: text

            git --version

Usage
~~~~~

> If you are not familiar with this tool and would like to know how to use it,
> please refer to the `Git Reference Manual <https://git-scm.com/doc>`_. In
> previous website, you will find plenty of resources for learning Git.

Reviewer: ```suggestion
If you are not familiar with Git, see the `Git Reference Manual <https://git-scm.com/doc>`_.
for comprehensive information on learning how to use it.
```

Configuration
~~~~~~~~~~~~~

> It is very important to properly configure Git so every modification you made to
> the code properly points towards you. There are two types of configuration:

Reviewer: ```suggestion
It is very important to properly configure Git so that every modification that you make
to the code points to you. There are two types of configuration:
```

> :ref:`Global` and :ref:`Local`. It is also possible to combine both and have

Reviewer: ```suggestion
:ref:`Global` and :ref:`Local`. It is also possible to combine both to have
```

:ref:`Multiple` configurations. 

Global
++++++

> Global configuration will be automatically included in ever Git repository of
> your machine unless override by a :ref:`Local` configuration. This configuration
> is located in ``C:\Users\<username>\.gitconfig`` for Windows users or in
> ``/home/<username>/.gitconfig`` for macOS/Linux/UNIX ones.

Reviewer: ```suggestion
Global configuration are automatically included in every Git repository on
your machine unless overridden by a :ref:`Local` configuration, which
is located in ``C:\Users\<username>\.gitconfig`` for Windows users or in
``/home/<username>/.gitconfig`` for macOS/Linux/UNIX users.
```

You can set the value for any variable in a field by running:

.. code-block:: bash

   git config --global <field>.<varname> <value>

> Some examples are provided in the lines below:

Reviewer: ```suggestion
Some examples follow.
```

**Setup your name:** 

.. code-block:: bash

    git config --global user.name <Your Name>

> **Setup your email:** 

Reviewer: ```suggestion
**Set up your email** 
```

.. code-block:: bash

    git config --global user.email <Ansys Email>

> **Setup default branch name:** 

Reviewer: ```suggestion
**Set up the default branch name** 
```

.. code-block:: bash

    git config --global init.defaultBranch main

Local
+++++
...

Multiple
++++++++
The dynamic configuration allows you to have a 

Signing Commits
~~~~~~~~~~~~~~~
.. todo::

Enabling SSH
~~~~~~~~~~~~
.. todo::

WSL2
----

> Some developers prefer using Windows as the operative system for their machines.
> However, they may like to take advantage from some features provided by a Linux
> system. To solve for this problem, the `Windows Subsystem for Linux
> <https://docs.microsoft.com/en-us/windows/wsl/install>`_ was devised.

Reviewer: ```suggestion
Some developers prefer using Windows as the operating system for their machines.
However, they may like to take advantage of some features provided by a Linux
operating system. The `Windows Subsystem for Linux
<https://docs.microsoft.com/en-us/windows/wsl/install>`_ was devised to solve
this problem.
```

Installation
~~~~~~~~~~~~

> Open a new PowerShell session and enter the following command:

Reviewer: ```suggestion
Open a new PowerShell session and run the following command:
```

.. code-block:: powershell

   wsl --install

> After installing ``WSL``, make sure you are running the ``WSL2`` version by
> executing:

Reviewer: ```suggestion
After installing ``WSL``, ensure that you are running the ``WSL2`` version with:
```

.. code-block:: powershell

   wsl --set-default-version 2

> Verifying Installation
> ~~~~~~~~~~~~~~~~~~~~~~
> To check which ``WSL`` version are you using, run:

Reviewer: ```suggestion
Verification
~~~~~~~~~~
To verify your ``WSL`` version, run:
```

.. code-block:: powershell

   wsl --list -v

> Installing a Linux Distro
> ~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Linux Distribution
~~~~~~~~~~~~~~
```

> Once ``WSL2`` has been installed, it is time to install a Linux
> distribution. To get a list of the available ones, run:

Reviewer: ```suggestion
After ``WSL2`` is installed, install a Linux distribution.
To get a list of available distributions, run:
```

.. code-block:: powershell

   wsl --list --online

> Most developers choose `Ubuntu <https://ubuntu.com/download>`_, as it is a
> heavily maintained Linux distribution with a huge collection of packages.

Reviewer: ```suggestion
Most developers choose `Ubuntu <https://ubuntu.com/download>`_ because it is a
well maintained Linux distribution with a huge collection of packages.
```

> Proceed to install the Linux distribution of your choice by running:

Reviewer: ```suggestion
To install the Linux distribution of your choice, run:
```

.. code-block:: powershell

>    wsl --install -d <distro name>

Reviewer: ```suggestion
   wsl --install -d <distribution name>
```

> You can install multiple Linux distributions by running previous command. To
> indicate ``WSL2`` which distribution you would like to use, execute:

Reviewer: ```suggestion
You can use this command to install multiple Linux distributions. To indicate
the distribution that you would like to use to ``WSL2``, run:
```

.. code-block:: powershell
   

>    wsl -d <distro name>

Reviewer: ```suggestion
   wsl -d <distribution name>
```

Windows Terminal
----------------

> The `Windows Terminal <https://docs.microsoft.com/en-us/windows/terminal/>`_ is
> an application devised to integrate in a single console multiple shells.
> Considering that Windows ships by default with two shells (``CMD`` and
> ``PowerShell``) and that installing :ref:`WSL2` adds a third one, ``Windows
> Terminal`` allows to collect all of them in a single program. 

Reviewer: ```suggestion
The `Windows Terminal <https://docs.microsoft.com/en-us/windows/terminal/>`_ is
an application that integrates multiple shells into a single console.
Considering that Windows ships by default with two shells (``CMD`` and
``PowerShell``) and that installing :ref:`WSL2` adds a third shell, ``Windows
Terminal`` provides for collecting them all in a single program. 
```

Installation
~~~~~~~~~~~~

> You can install `Windows Terminal from Microsoft Store

Reviewer: ```suggestion
You can install `Windows Terminal
```
